### PR TITLE
[quantization] Support Evaluation of Qwen3-VL With MMMU Dataset

### PR DESCRIPTION
**Pull request by dvsav**
# What

Adds support for evaluating Qwen3-VL models on the MMMU (Massive Multi-discipline Multimodal Understanding) benchmark dataset.

# Why

Addresses [PTQ Evaluation — Qwen3-VL](https://github.sec.samsung.net/one-project/TICO/issues/192) issue which requested MMMU dataset support for comprehensive VLM evaluation.

# Implementation Details

**New File:**
- `tico/quantization/evaluation/mmmu_eval_utils.py` — MMMU evaluation utilities including:
  - Dataset loading and streaming for 30 MMMU subjects
  - Few-shot prompt construction with multichoice formatting
  - Answer extraction (A/B/C/D) from model output
  - Per-subject and aggregate evaluation functions
  - Reuses common utilities from `vlm_eval_utils.py`

**Modified File:**
- `tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py`:
  - Added CLI arguments: `--mmmu_subjects`, `--mmmu_n_shots`, `--mmmu_n_samples`
  - Integrated MMMU evaluation for both original and quantized models

# Example
```bash
python tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py \
    --model=Qwen/Qwen3-VL-4B-Instruct \
    --cache_dir=/home/d.savchenkov/models/qwen3-vl-4b \
    --trust-remote-code \
    --mmmu_subjects Art Geography Biology \
    --gptq_mse=mse \
    --no_GPTQ \
    --mmmu_n_samples=10 \
    --embedding_weight_bits=16 \
    --vision_patch_embed_weight_bits=16 \
    --linear_weight_bits=16 \
    --lm_head_weight_bits=16 \
    --nsamples_for_qcalibration=10

=== MMMU Evaluation (Original Model) ===
| subject                                            | correct    | total      | accuracy   |
| -------------------------------------------------- | ---------- | ---------- | ---------- |
| Art                                                | 5          | 10         | 0.5000     |
| Geography                                          | 1          | 10         | 0.1000     |
| Biology                                            | 6          | 10         | 0.6000     |

=== MMMU Evaluation (Quantized Model) ===
| subject                                            | correct    | total      | accuracy   |
| -------------------------------------------------- | ---------- | ---------- | ---------- |
| Art                                                | 5          | 10         | 0.5000     |
| Geography                                          | 1          | 10         | 0.1000     |
| Biology                                            | 6          | 10         | 0.6000     |
```

### CODE REVIEW

File: tico/quantization/evaluation/mmmu_eval_utils.py [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+    if answer_match:
+        return answer_match.group(1).upper()
+
+    # Strategy 2: Look for standalone letter at the beginning
+    first_char_match = re.match(r"^([A-D])[.\s]", text, re.IGNORECASE)
+    if first_char_match:
+        return first_char_match.group(1).upper()
+
+    # Strategy 3: Look for first occurrence of A/B/C/D
+    for char in text:
```
Comment (Torrero) [inline tico/quantization/evaluation/mmmu_eval_utils.py]: I have some doubts about this strategy. 
In case the model produces some irrelevant answer, we increase the chances of the model providing right answer by this strategy. 
Is it standard strategy for evaluation this dataset?

File: tico/quantization/evaluation/mmmu_eval_utils.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+import torch
+from datasets import Dataset, IterableDataset, load_dataset
+
+from tico.quantization.evaluation.vlm_eval_utils import (
+    build_messages,
+    build_prompt,
+    build_vlm_inputs,
+    generate_answer,
+    move_inputs_to_device,
+)
```
Comment (Torrero) [inline tico/quantization/evaluation/mmmu_eval_utils.py]: There are some unused import, could check please

File: tico/quantization/evaluation/mmmu_eval_utils.py [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+        subject=subject, start=n_shots, n_samples=n_samples, streaming=True
+    )
+
+    correct = 0
+    total = 0
+
+    for ex in test_data:
+        # Skip questions with multiple images
+        if ex["image_2"] is not None:
+            continue
```
Comment (Torrero) [inline tico/quantization/evaluation/mmmu_eval_utils.py:291]: (optional) Maybe here we can keep statistics of skipped samples and also print it further

Comment (dvsav) [reply tico/quantization/evaluation/mmmu_eval_utils.py:291]: 👍 done
Now the output table looks like this:
```
| subject                                            | correct    | total      | skipped    | accuracy   |
| -------------------------------------------------- | ---------- | ---------- | ---------- | ---------- |
| Chemistry                                          | 1          | 8          | 2          | 0.1250     |
| History                                            | 3          | 9          | 1          | 0.3333     |
```

File: tico/quantization/evaluation/mmmu_eval_utils.py (hunk 2 above)
Comment (dvsav) [reply tico/quantization/evaluation/mmmu_eval_utils.py]: 👍 done

File: tico/quantization/evaluation/mmmu_eval_utils.py (hunk 1 above)
Comment (dvsav) [reply tico/quantization/evaluation/mmmu_eval_utils.py]: 👍 I've reduced the extraction to a single regex that (among others) handles patterns like `A. 4` that I saw in a real test:
```
[Sample 7] Subject: Chemistry
Q: How many chiral center (excluding N centres) are there in morphine? <image 1>...
Choices: ['4', '5', '6', 'More than 6']
Generated: A. 4, Predicted: A, Gold: B, Correct: False
```

**Review by Torrero — DISMISSED** [review]
LGTM, thank you

File: tico/quantization/evaluation/mmmu_eval_utils.py [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+        max_new_tokens: Maximum tokens to generate.
+        temperature: Sampling temperature.
+        verbose: Whether to print detailed logs.
+
+    Returns:
+        A tuple of (correct_count, total_count, skipped_count).
+    """
+    few_shot_examples = load_few_shot_examples(subject=subject, n_shots=n_shots)
+
+    test_data = load_data(
```
Comment (mhs4670go) [inline tico/quantization/evaluation/mmmu_eval_utils.py:272]: 5-shot examples should be included in the prompt, but they should not be evaluated again as target samples. 

IIUC, currently `evaluate_subject()` passes `start=n_shots`, but `load_data()` ignores start when `n_samples=-1`, which is the default full-evaluation mode. As a result, the first `n_shots` validation examples can be used both as few-shot examples and as evaluation targets.

Alternatively, we can use `dev` for few-shot examples and `validation` for target evaluation, which avoids overlap more cleanly.

File: tico/quantization/evaluation/mmmu_eval_utils.py [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+
+    Args:
+        subject: The subject name.
+        n_shots: Number of few-shot examples to load.
+
+    Returns:
+        List of example dictionaries with 'question', 'choices', and 'answer'.
+    """
+    ds = load_data(
+        subject=subject,
```
Comment (mhs4670go) [inline tico/quantization/evaluation/mmmu_eval_utils.py:232]: (Optional) `n_shots=0` currently does not behave as zero-shot. Because `load_data()` only truncates when `n_samples > 0`, `n_samples=0` returns the full split, so `load_few_shot_examples(..., n_shots=0)` may load all validation examples as few-shot examples. Please add an early return for zero-shot, e.g. if n_shots <= 0: return [].

```suggestion
    if n_shots <= 0:
        return []
        
    ds = load_data(
        subject=subject,
```

File: tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py [hunk 6]
```diff
@@ -322,6 +326,31 @@ def parse_args():
...
 
+    # MMMU evaluation arguments
+    parser.add_argument(
+        "--mmmu_subjects",
+        type=str,
+        default=None,
+        nargs="+",
+        help=(
+            "Space-separated list of MMMU subjects to evaluate. Use 'mmmu' for all subjects."
+            "Use 'accounting', 'agriculture', 'art', etc. dpecific subjects."
```
Comment (mhs4670go) [inline tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py]: User may mistakenly assume that the subjects are case-insensitive.

```suggestion
            "Use 'Accounting', 'Agriculture', 'Art', etc. dpecific subjects."
```

Comment (dvsav) [reply tico/quantization/wrapq/examples/quantize_qwen3_vl_with_gptq.py]: 👍 done

File: tico/quantization/evaluation/mmmu_eval_utils.py (hunk 5 above)
Comment (dvsav) [reply tico/quantization/evaluation/mmmu_eval_utils.py:232]: 👍 done

File: tico/quantization/evaluation/mmmu_eval_utils.py (hunk 4 above)
Comment (dvsav) [reply tico/quantization/evaluation/mmmu_eval_utils.py:272]: 👍 Good point, thank you!
I've switched `load_few_shot_examples` to using "dev" split and also fixed `take_from_dataset` so that it handles `n <= 0`.

**Review by mhs4670go — APPROVED** [review]
LGTM
